File: src/main/java/com/example/librarymanagement/repository/BookPresenceRepository.java
package com.example.librarymanagement.repository;

import com.example.librarymanagement.model.entity.BookPresence;
import com.example.librarymanagement.model.enums.Availability;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;

public interface BookPresenceRepository extends JpaRepository<BookPresence, Long> {
    List<BookPresence> findByLibraryId(Long libraryId);
    List<BookPresence> findAllByBookId(Long bookId);
    List<BookPresence> findAllByLibraryId(Long libraryId);
    List<BookPresence> findAllByUserId(Long userId);
    List<BookPresence> findAllByLibraryIdAndBookId(Long libraryId, Long bookId);
    List<BookPresence> findAllByLibraryIdAndBookIdAndAvailability(Long libraryId, Long bookId, Availability availability);
}

File: src/main/java/com/example/librarymanagement/repository/BookPresenceRepository.kt
package com.example.librarymanagement.repository

import com.example.librarymanagement.model.domain.BookPresence
import com.example.librarymanagement.model.domain.User
import com.example.librarymanagement.model.enums.Availability

interface BookPresenceRepository {
    fun save(bookPresence: BookPresence): BookPresence
    fun addBookToUser(user: User, libraryId: String, bookId: String): BookPresence?
    fun removeBookFromUser(user: User, libraryId: String, bookId: String): BookPresence?
    fun findById(bookPresenceId: String): BookPresence?
    fun deleteById(bookPresenceId: String)
    fun findAllByBookId(bookId: String): List<BookPresence>
    fun findAllByLibraryId(libraryId: String): List<BookPresence>
    fun findAllByUserId(userId: String): List<BookPresence>
    fun findAllByLibraryIdAndBookId(
        libraryId: String,
        bookId: String
    ): List<BookPresence>

    fun findAllByLibraryIdAndBookIdAndAvailability(
        libraryId: String,
        bookId: String,
        availability: Availability
    ): List<BookPresence>

    fun findAllByLibraryIdAndAvailability(libraryId: String, availability: Availability): List<BookPresence>
    fun existsByBookIdAndLibraryId(bookId: String, libraryId: String): Boolean

}
